addons/website_slides_survey/models/survey_survey.py

```diff
@@ -32,7 +33,14 @@ def _unlink_except_linked_to_course(self):
         # even if they don't have access to those surveys hence the sudo usage
         certifications = self.sudo().slide_ids.filtered(lambda slide: slide.slide_type == "certification").mapped('survey_id').exists()
         if certifications:
-            certifications_course_mapping = [_('- %s (Courses - %s)', certi.title, '; '.join(certi.slide_channel_ids.mapped('name'))) for certi in certifications]
+            certifications_course_mapping = [
+                _(
+                    "- %(certification)s (Courses - %(courses)s)",
+                    certification=certi.title,
+                    courses=format_list(self.env, certi.slide_channel_ids.mapped("name")),
+                )
+                for certi in certifications
+            ]
```

Comment:
Cloud you double check if it really works?
the code translation `_` uses the variable e.g. `context` `self.env.lang` in the frame to determine the language to translate, but list generator has its own frame. Usually there is nothing to infer the `lang`. And it should always be English.
But here you are lucky, bucause `self` is also used in the generator and can be found in the frame.

I suggest we translate outside of the generator to avoid the tricky behavior as the best pratice

Reply:
@HydrionBurst I did a small refactoring in my Context API PR (that is on hold) here: https://github.com/odoo/odoo/pull/155765/files#diff-9f9f8f7fdb4edf151cf1ba03efc0e5dd7ea6857d431d5a6f80fe769977ac7161L524-R562

Might that be a solution, so it doesn't matter how many stackframes are in between the call to _() and the translation code?

Reply:
I don't think it is the solution.
The problem is the generator/list comprehension/local function has its own frame
```py
def method_name(self):
    _('src1')
    return `[_('src2') for i in range(3)]`
```
`_('src1')` and `_('src2')` are in different frames with the same `inspect.getfile(frame)`
you can get `self` from the `f_locals` of the frame of `_('src1')`
but you cannot get `self` from the `f_locals` of the frame of `_('src2')`

But if you write the code like this
```py
def method_name(self):
    _('src1')
    return `[(_('src2'), bool(self)) for i in range(3)]`
```
you can get `self` from both `f_locals` of the frames of `_('src1')` and `_('src2')`
This is because `self` is also used in the frame of the list comprehension `[(_('src2'), bool(self)) for i in range(3)]`

If we really want to improve the developer experience
we can do another f_back loop
```py
            while (model_file := inspect.getfile(frame)) == translate_file:
                frame = frame.f_back
            while True:
                locals = frame.f_locals
                if 'self' in locals or 'context' in locals or 'context' in locals.get('kwargs', {}):
                    break
                frame_back = frame.f_back
                if inspect.getfile(frame_back) != model_file:
                    break
                frame = frame_back
            lang = self._get_lang(frame)
......
```

Reply:
You're right, I mixed it up. Your last suggestion here seems pretty nice though and would indeed enhance the developer experience, I think.